### ISSUE TITLE
Idea: new cannon types

### ISSUE THREAD

**Issue by kiedtl** [pandorabox-io/spacecannon#15]
I'm experimenting with creating two new "railgun" cannon types that possess the following properties:

- Require "ammunition" in the form of special "slug" items.
- Fire extremely fast projectiles, though not so fast that they cannot be dodged by an agile ship/pilot.
- Cannons require very little energy and practically no charging period, allowing for rapid-fire shots.
- Unsure if possible: cannons can only hold, say, 10-20 ammo items (instead of a whole stack), to ensure rapid-fire shooting isn't overpowered.
- Projectiles make no explosion, instead penetrating multiple layers and destroying a single node.
- Projectiles also have a much longer range compared to the regular cannons.

The idea is that these cannons fire kinetic projectiles instead of whatever energy-based thing the current cannons use, hence the need for ammo and the reduced need for energy.

The base railgun would require maybe 300 EU (and store just that), while penetrating two layers of the target.

The "helical railgun" would be an upgraded version of the railgun, that penetrates 4 layers of the target while having a longer range and taking up more energy.

Currently I'm adding these cannons purely for my enjoyment, but if it sounds like something that might fit this mod I can make a PR and maybe discuss other changes

**Comment by BuckarooBanzay** [comment]
> Currently I'm adding these cannons purely for my enjoyment, but if it sounds like something that might fit this mod I can make a PR and maybe discuss other changes

Sure :+1: every snippet of code/image would help in that case

**Comment by S-S-X** [comment]
Projectile requirement sounds good, I think that could also allow bit overpowered fast explosive projectiles if manufacturing requires a lot of resources and possibly multiple steps.

Simple slugs could well be just piece of metal while way more powerful explosive projectiles could require depleted uranium (which already requires high tech manufacturing plant and does not yet have many uses), a lot of electronics (just use some random stuff from other mods), probably some plastic and various shaped metal parts.

I guess that with limited inventory size for auto loader would be well enough to allow some very powerful weapons (both fast and with large explosion).

**Comment by kiedtl** [comment]
In my current PR slugs consist of two carbon steel strips, four stainless steel ingots, and one depleted uranium ingot. Not extremely expensive, but expensive enough that it can't be mass-produced in the hundreds of stacks.

Still, I like the idea of making them more expensive and more difficult to produce.

About using electronics in slug crafting, it seems counterintuitive since the whole point of a railgun is that you fire a dumb piece of metal at high speeds, that is incapable of maneuvering or independent targeting. It's not a missile after all.

I feel like different types of slugs, explosions, and such would fit better in a new "missile launcher" cannon with its own set of unique properties. That way one could have multiple kinds of missiles, some extremely expensive and some not so much, some capable of 45 degree targeting and others dumb pieces of steel, some with big explosions and others with smaller ones.

**Comment by S-S-X** [comment]
> About using electronics in slug crafting, it seems counterintuitive

Note that I was talking about explosive rounds, not simple slugs. For me it sounds that slugs are fine with cheap materials but possible explosive high speed rail gun projectiles are not. They'd need way more complicated crafting requirements in addition to limited auto loader inventory (like 5 rounds per cannon instead of 99 stacks, if there would be need to prevent scripted clients doing reload then server side timer for reloading 100% prevents that).

So basically I was talking about over powered high tech high speed explosive armor piercing bunker buster projectiles that are not rockets but rather grenade like rail launched direct fire munition.

I'm not aware of such things in real world but basically what I was thinking would be basically rail gun with real armor penetration abilities, not just what is delivered through kinetic energy.

So yeah, missile launcher but without missiles :p

**Comment by kiedtl** [comment]
I'm going to dump a list of stats here which future slug items could modify, for future reference. (I'll update as I remember.)

- Damage intensity (for armor piercing rounds)
- Explosion radius, if any
- How many explosions, and where (at the end? at the beginning? all along the penetration route?)
- Number of projectiles (potential for "multislug" item that causes cannon to fire several rounds at once, each at a slight different angle for maximum damage. aka "flak cannon")
- Damage type (perhaps we could have slugs doing electromagnetic damage that scrambles/disables mesecon/digilines?)
- Homing (slugs that chase down targets that jumped out of the projectile path!)
- Overloading (where surplus cannon energy translates to extra damage/range/etc)

**Comment by kiedtl** [comment]
Dumping some prior discussions w/ @.dennisjenkins75

> [05/25/2023 1:26 PM] 1hittoaster#4196 railguns typically have a huge capacitor bank, as they need massive current (1M+ amps).  Maybe add such a capacitor node (hv_capacitor).  its like a battery, in that it can hold EUs.  It will leak them if not continually charged, is expsnsive to build, and the railgun will check its HV network to see that there are sufficient capacitors available with enough total charge before firing, and then deplete the capacitor.
> [05/25/2023 1:26 PM] 1hittoaster#4196 So the cost is in the PSU, not the projectiles.
> [05/25/2023 1:27 PM] 1hittoaster#4196 The caps would need compressed copper plates, graphite, oil, plastic.
> [05/25/2023 1:27 PM] 1hittoaster#4196 Or "compressed mithril plates"
> [05/25/2023 1:27 PM] 1hittoaster#4196 or something exotic.
> <snip>
> [05/25/2023 1:28 PM] 1hittoaster#4196 or have the whole stack: lv_capacitor, mv_cap, hv_cap.
> [05/25/2023 1:28 PM] 1hittoaster#4196 lv_cap is expensive, and you need 5 to make am mv_cap
> [05/25/2023 1:28 PM] 1hittoaster#4196 and 5 mv_caps to make an hv_cap
> [05/25/2023 1:28 PM] 1hittoaster#4196 and 10 hv_caps to fire the rail gun.
> [05/25/2023 1:29 PM] 1hittoaster#4196 the problem w/ making it uber expsnsive, is that quite a few players can still mass-produce them anyway.
> [05/25/2023 1:29 PM] 1hittoaster#4196 IRL, most railgun barrels warp with usage, and fail after only a few firings, and need to be reconditioned.
> [05/25/2023 1:29 PM] 1hittoaster#4196 IDK how much realism you want to go for.
> [05/25/2023 1:31 PM] kiedtl#5513 Yep, that's the core issue I guess. I don't want to make it so expensive that it's an elite weapon that needs years of grinding to be able to use.
> [05/25/2023 1:32 PM] kiedtl#5513 Making the railgun itself "burn out" after, maybe, 30-40 shots sound really interesting though
> [05/25/2023 1:33 PM] kiedtl#5513 Unsure if it's possible, but perhaps it could simply be placed into a tool workshop or something similar to fix -- sure it could be automated, but that's fine given you'd then have to be stuffing a whole MV+workshop network inside your ship
> [05/25/2023 1:48 PM] 1hittoaster#4196 I would hate it burning out after usage, but that's more realistic.
> [05/25/2023 1:48 PM] 1hittoaster#4196 ppl could automate the replacement w/ a digi-device anyway
> [05/25/2023 1:50 PM] 1hittoaster#4196 Consider ammending your proposal to include the capacitor idea.  They hold small-ish amounts of EU, leak it over time if not continually charged up, and the rail gun requires that they be connected to the same HV network as the railgun, and be reasonably close (within a 15 nodes?) and have a total charge required to fire the railgun, and recharge from the local HV network slowly.
> [05/25/2023 1:59 PM] kiedtl#5513 I'll mention it in the issue, but it doesn't really appeal to me much since 1) it's a one time cost and 2) it negates one of the intended benefits of railguns, fast fire times
> [05/25/2023 2:07 PM] 1hittoaster#4196 ok, np.  Was just an idea.
> [05/25/2023 2:07 PM] 1hittoaster#4196 the capacitors can charge at whatever speed you want.
> [05/25/2023 2:07 PM] 1hittoaster#4196 But I understand, its massive scope creep.

**Comment by S-S-X** [comment]
Capacitors could provide other uses as well and been discussed before. However currently not many direct use cases.
Anyway, if added, I would like to make capacitor generic Technic component instead of space cannon specific thing and just make space cannon power requirement very high.

Very large Technic battery array could also provide similar output capacity, currently max short circuit power / hv battery is 400kEU so for example 4 batteries can provide max 1600kEU output + in addition to any connected generator output.
Like suggested, capacitors could have a lot higher short circuit power but leak over time which wouldn't really be too hard to implement and is actually very similar to regular Technic battery functionality save leaking which should be fairly simple to add.

**Comment by kiedtl** [comment]
Whew... one year later. Seems like 10 years ago.

I wish I had written a design note about what I want from this, just for future reference. Coming to this now, I don't understand what the point was with rapid-fire. I also haven't played the game since then, truth be told.

One benefit (that I've just thought of from rereading this thread) of requiring a capacitor could be that it needs a larger space than a regular cannon. That immediately adds an interesting design factor, and a one-time cost, without making it something most people can't afford but a few high-level players can crank out in massive quantities.

I've also come around to the idea of the cannon warping. Maybe the warping could have other effects, such as causing the projectile to fire at an angle?
